### PR TITLE
Fix bug app goes back to MainActivity when user clicks home while on NewsDetailActivity

### LINKED ISSUE #73: NewsDetailActivity is Destroyed Unintentionally
From one of our user:

```
The 1 is that when I have the app open to an article's comments,
hit home,
go to recent apps and open it,
it puts me back in the list of articles rather the comments I was just looking at.
I don't think it's a an issue with ram but I could be wrong.
```

This one is caused by **circular reference** on model Comment (it has parent comment). By default, Parceler doesn't handle it. When it tries to wrap an object that contains circular reference, it throws StackOverflowError. This caused the NewsDetailActivity (that contains NewsDetailFragment) to be destroyed and app is back to MainActivity.

### PR DESCRIPTION
**Pull request by timotiusnc**
Fixes #73 

This is really weird.
The troublesome line of codes is this:

```
@Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);

        if (newsDetailAdapter != null) {
            outState.putParcelable(SAVED_DATASET, Parcels.wrap(newsDetailAdapter.getDataSet()));
            outState.putParcelable(SAVED_COMMENTS, Parcels.wrap(newsDetailAdapter.getComments()));
        }
    }
```

on `NewsDetailFragment.java`
It returns StackOverflow error because we try to wrap model with circular reference. In this case: Comment has `childComments` and `parentComment` attributes.

When there's rotation changes or user hits home (or lock the screen), that `onSaveInstanceState` is called. The weird thing is:
- if rotation changes, the method works fine (Parceler can wrap it)
- if Home button is hit or screen is locked, the method throws StackOverflowError

My first thought for solution is to create a very special way to save the instance:
1. Declare another Comment model that contains every original Comment model attributes **except** `List<Comment> childComments` and `Comment parentComment`.
2. Save those new Comment model instead (should be working because there's no circular reference anymore).
3. When unwrapping, populate that `List<Comment> childComments` and `Comment parentComment` using recursive method.

But then, an even weirder thing happened. I tried to implement Parcelable on model Comment first:

```
@Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(android.os.Parcel dest, int flags) {
        Log.d("Pumpkin", "writeToParcel");
        dest.writeInt(parent);
        dest.writeList(kids);
        dest.writeByte((byte) (expanded ? 1 : 0));
        dest.writeByte((byte) (hidden ? 1 : 0));
        dest.writeInt(level);
        dest.writeInt(allChildCount);
    }
```

I still haven't handled the recursive part, then I decided to try it out first. My expectation is that **writeToParcel** method will be called on `onSaveInstanceState` method.

It turned out: the **writeToParcel** method above only got called when user hits home (or lock the screen). It doesn't get called if user changes rotation. So, only implementing Parcelable actually save the problem!

Then I tried another thing: I removed the `@Parcel` annotation on model Comment (but still implementing Parcelable). My expectation is that the `onSaveInstanceState` will still work (because we still implement Parcelable). It turned out: when user changes screen orientation: yes it still works BUT the `writeToParcel` method is NOT called!

...and what about when user hit home screen? It throws error:

```
 Unable to find generated Parcelable class for io.pumpkinz.pumpkinreader.model.Comment,
verify that your class is configured properly and that the Parcelable class
io.pumpkinz.pumpkinreader.model.Comment$$Parcelable is generated by Parceler.
```

WTF!?!?!?

Actually, based on the documentation on http://developer.android.com/guide/components/activities.html#Lifecycle. Only relying on this solution is not robust.

When user hits home or lock the screen, the activity is on Paused state (hence, the activity is not destroyed). When user comes back to app, the app just resumes, that's why **although I haven't handled that recursive part (populating childComments and parentComment)**, it will still work.

But, when on Paused state and there's another app with higher priority requires memory. The paused activity will be killed, and when user open it again, it will be recreated, this time, the state will be different because the comments don't have child or parent.

What do you say @wiseodd ?

### CODE REVIEW

**Comment by timotiusnc** [comment]
Decided to go with the robust way. If you want to check the thing I mentioned before, you check it out on first commit of this PR.

**Comment by wiseodd** [comment]
Wait, I'm kinda lost it.
So, what is "the robust solution"?

**Comment by timotiusnc** [comment]
Hemm...now that I have re-read the first post, I am not sure myself if I would understand it in the future. Haha. OK, here's the more concise version.

**Problem I:**
Parceler (org.parceler) could not process circular reference on `Comment` model (it causes stack overflow error).

**Problem II:**
`onSaveInstanceState` method is called when user changes rotation and hits home button or lock the screen. Inside `onSaveInstanceState`, app calls `Parcels.wrap` method to serialize `List<Comment`>.
`Parcels.wrap` behaves in weird way. It throws StackOverflow error when `onSaveInstanceState` method is called when user hits home or lock the screen. It does NOT throw StackOverflow error when `onSaveInstaceState` method is called when user changes rotation (this is why we missed it, we tested it only on rotation changes, not hit home button or screen lock).

**Problem III:**
The `implements Parcelable` (android.os.Parcelable) also behaves in weird way. Just like problem II, the `writeToParcel` method is called when user hits home or lock the screen. It is NOT called when user changes rotation.

**Solution I (the not robust):**
Still use `@Parcel` annotation in Comment model AND also `implements Parcelable`. This is like taking advantage of that weird behavior (just so we won't get StackOverflow error). This is not robust because the information of child and parent comment is not serialized. Based on Android activity and fragment lifecycle documentation, we will hit problem in a very rare occasion when the device is running low on memory.

**Solution II (the robust):**
Omit `@Parcel` annotation and omit (also) `implement Parcelable` on Comment model. Introduce new helper class to do the serialization, that is `CommentParcel` class which has all the attributes of Comment model except the circular reference ones. We can do this because the info about the circular reference attributes (child and parent comment) can be inferred on the fly (that recursive method).

PS: I'm not really sure with the code on `CommentParcel` though. It looks ugly (lots of `static` methods).

**Comment by wiseodd** [comment]
That's one perspective for looking at this problem.
Another is to thinker about our comment data structure.

What if we create non recursive data structure for comment?
Like, we still have methods like getParent(), but instead of pointing to other comment object directly, which will cause the circular reference, we calculate it instead.

For example, if we make the comments as list of comments and parent-child relationship is signified by comments' level (1, 2, 3, ...), then in our getParent(), we can do a calculation: "given this comment object with level x, what is the parent?"

Something like that.

What do you think?

**Comment by timotiusnc** [comment]
That's what I was thinking before implementing the helper, I thought it will be much cleaner to serialize the comment **without** the circular reference part, but when we deserialize it, we calculate the circular reference part.

But then I was stuck with: in order to calculate the parent and children comment, I need to know all other comments, whereas we're talking about calculating them on only one comment point of view.

In a nutshell, the question is: "How can I calculate the parent and children if I don't know the rest of the comments?"

Other way to "calculate" it is not calculating it in Comment model, object that needs parent will need to have all comments (that non-recursive list of comments) and call `getParentId()` on a comment, then lookup that ID.

**Comment by wiseodd** [comment]
I've tested this, and it seems the bug has been fixed.
I'll just accept this code though for now, merely to make the release faster.
I'm in no condition to review this now.

**Comment by wiseodd** [comment]
By the way, please don't delete this branch, as I'll visit this later.
